tests/AstrolabeWorkloadExecutor/Program.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.IO;
+using MongoDB.Bson;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson.IO;
+
+namespace WorkloadExecutor
+{
+    class Program
+    {
+        static long __numberOfSuccessfulOperations;
+        static long __numberOfFailedOperations;
+        static long __numberOfOperationErrors;
+
+        static System.Diagnostics.Stopwatch __stopwatch = new System.Diagnostics.Stopwatch();
+        static async Task Main(string[] args)
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+            ConsoleCancelEventHandler cancelHandler = (o, e) => HandleCancel(e, cancellationTokenSource);
+
+            var resultsDir = Environment.GetEnvironmentVariable("RESULTS_DIR");
+            var resultsPath = resultsDir == null ? "results.json" : Path.Combine(resultsDir, "results.json");
+            Console.WriteLine($"dotnet main> Results will be written to {resultsPath}..." );
+
+            try
+            {
+                // register all interrupt and exit handlers
+                Console.CancelKeyPress += cancelHandler;
+                foreach (var arg in args)
+                {
+                    Console.WriteLine($"dotnet main> args: {arg}");
+                }
+
+                var connectionString = args[0];
+                var driverWorkload = BsonDocument.Parse(args[1]);
+                Console.WriteLine("dotnet main> Starting workload executor...");
+
+                if (!bool.TryParse(Environment.GetEnvironmentVariable("ASYNC"), out bool async))
+                {
+                    async = true;
+                }
+                if (async)
+                {
+                    await ExecuteWorkloadAsync(connectionString, driverWorkload, cancellationTokenSource.Token).ConfigureAwait(false);
+                }
+                else
+                {
+                    ExecuteWorkload(connectionString, driverWorkload, cancellationTokenSource.Token);
+                }
+
+
+            }
+            finally // HandleCancel and this finally clause must finish executing well under 50ms if we want them to finish
```

Comment:
Later tests runs showed that sometimes we'll have as much as 100ms, so I will update this comment so we don't lose this particular piece of information